File: Array/subarray/subarray_sum_equals_k.kt
package Array.subarray


fun main() {
    var nums = intArrayOf(1, 1, 1)
    var k = 2
    println(subarraySum(nums, 2))
}

// Subarray equals sums k in leetcode
/*

Given an array of integers nums and an integer k,
 return the total number of subarrays whose sum equals to k.
 Input: nums = [1,2,3], k = 3
Output: 2
 */
// Need to work on the prefix sum problem here
fun subarraySum(nums: IntArray, k: Int): Int {

    /*
    3   4   7    2          k = 7
 0  3   7   14   16         we have to add the sum here

 @ each iteration then we have 7 -7 = 0, so that means we found 1, then we add that to the
 code or our dict here
     */
    // See the notes
    var dict = mutableMapOf<Int, Int>()

    // This is the part of the trick here too
    dict[0] = 1
    var n = nums.size
    var count = 0
    var s = 0

    for (num in nums) {
        s += num
        // The case where 14 - 7 is already in there
        if (dict.contains(s - k)) {
            dict[s - k]?.let {
                count += it
            }
        }

        // The case 2 where if already seen the sum 14
        if (dict.contains(s)) {
            dict[s] = dict[s]!! + 1
        }
        // The case if havent seen before here
        else {
            dict[s] = 1
        }

    }
    return count
}

File: Array/subarray/subarray_sum_equals_k.py

'''

How to deal with this problem here?

And then using this we have the code
- use a hashmpap

to find the sum that we see along the way here

Find the # of times that

1. This is key here:
we attempt to find the number of times the complement of the current cumulative sum,
(s - k), has already appeared.

2.

'''
from collections import Counter, defaultdict
from typing import List


def subarraySum(self, nums: List[int], k: int) -> int:
    # Initialize a counter to keep track of the cumulative sums encountered
    cumulative_sum_counter = Counter({0: 1})

    # 'count_subarrays' will store the total count of subarrays that sum up to 'k'
    count_subarrays = 0

    # 'cumulative_sum' holds the sum of numbers seen so far
    cumulative_sum = 0

    # Iterate through the list of numbers
    for num in nums:
        # Update the cumulative sum
        cumulative_sum += num

        # If there is a previous cumulative sum such that current_sum - k
        # is equal to that previous sum, then a subarray ending at the current
        # position would sum to 'k'
        count_subarrays += cumulative_sum_counter[cumulative_sum - k]

        # Increase the count of the current cumulative sum by 1 in the counter
        cumulative_sum_counter[cumulative_sum] += 1

    # Return the total number of subarrays that sum up to 'k'
    return count_subarrays
